### PR TITLE
2 browse from tree

### LINKED ISSUE #2: Als Squeak-Entwickler*in möchte ich direkt aus dem Profilingtool die Methoden im Browser öffnen können um sie zu bearbeiten.

### PR DESCRIPTION
**Pull request by linusha**
this closes #2.

### CODE REVIEW

File: packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/currentSelection..st [hunk 1]
```diff
@@ -0,0 +1,6 @@
+accessing
+currentSelection: anObject
```
Comment (twollnik) [inline packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/currentSelection..st]: I think we can be more precise in the naming of the argument. What do you think of `aNode`?

Comment (linusha) [reply packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/currentSelection..st]: Agreed and changed.

File: packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/browseMethodCall.st [hunk 2]
```diff
@@ -0,0 +1,12 @@
+as yet unclassified
```
Comment (tillprochaska) [inline packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/browseMethodCall.st]: This method is missing a category.

File: packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/explorerKey.from.event..st [hunk 3]
```diff
@@ -0,0 +1,9 @@
+key handling
+explorerKey: aChar from: view event: event
+
+	event anyModifierKeyPressed ifFalse: [^ false].
+
+	self currentSelection ifNotNil: [
```
Comment (tillprochaska) [inline packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/explorerKey.from.event..st]: Ich persönlich finde es einfacher, den Sinn zu erkennen, wenn die Bedingung tatsächlich als Konjunktion und nicht verschachtelt dargestellt wird.

```smalltalk
(self currentSelection notNil) & (aChar == $b)
	ifTrue: [self browseMethodCall].
```

Oder gleich mit early exit:

```smalltalk
self currentSelection ifNil: [^ false].
aChar == $b ifTrue: [self browseMethodCall].
```

File: packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/explorerKey.from.event..st [hunk 4]
```diff
@@ -0,0 +1,9 @@
+key handling
+explorerKey: aChar from: view event: event
```
Comment (tillprochaska) [inline packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/explorerKey.from.event..st]: Is there a reasons why this method has the term explorer in its name? Personally I think smth like `handleKey: aChar from: view event: event` would be more intuitive in this case.

File: packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/mainTreeListMenu..st [hunk 5]
```diff
@@ -0,0 +1,10 @@
+as yet unclassified
+mainTreeListMenu: aMenu
+	<treeListMenu>
```
Comment (tillprochaska) [inline packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/mainTreeListMenu..st]: Could you explain this syntax to me? Didn’t see it before and tried to find out about it online, but without luck… 😢

File: packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/mainTreeListMenu..st [hunk 6]
```diff
@@ -0,0 +1,10 @@
+as yet unclassified
```
Comment (tillprochaska) [inline packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/mainTreeListMenu..st]: This misses a category as well.

File: packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/mainTreeListMenu..st [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+mainTreeListMenu: aMenu
+	<treeListMenu>
+	
+	aMenu addStayUpItemSpecial.
```
Comment (tillprochaska) [inline packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/mainTreeListMenu..st]: What does `addStayUpItemSpecial` do? For me, it doesn’t seem to make any difference at all. And to be honest, I didn’t really understand the method comment in `MenuMorph`…

File: packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/mainTreeListMenu..st [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+
+	aMenu addList: #(
+		"TODO: add keyboard shortcut"
```
Comment (tillprochaska) [inline packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/mainTreeListMenu..st]:
```suggestion
```

Der Kommentar ist nicht mehr aktuell, oder? Ihr habt ja schließlich schon einen Shortcut implementiert.

File: packages/InteractiveProfilingTool-Tests.package/IPTTreeModelMock.class/class/computeSomething.st [hunk 9]
```diff
@@ -0,0 +1,5 @@
+as yet unclassified
+computeSomething
```
Comment (tillprochaska) [inline packages/InteractiveProfilingTool-Tests.package/IPTTreeModelMock.class/class/computeSomething.st]: Shouldn’t this method be part of `IPTTreeModelTestCase`?

File: packages/InteractiveProfilingTool-Tests.package/IPTTreeModelMock.class/class/constructMessageTallyFrom..st [hunk 10]
```diff
@@ -0,0 +1,9 @@
+as yet unclassified
+constructMessageTallyFrom: aBlockClosure
```
Comment (tillprochaska) [inline packages/InteractiveProfilingTool-Tests.package/IPTTreeModelMock.class/class/constructMessageTallyFrom..st]: Same as above, shouldn’t this be part of `IPTTreeModelTestCase`?

File: packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testBrowseClassMethod.st [hunk 11]
```diff
@@ -0,0 +1,16 @@
+tests
+testBrowseClassMethod
+	
+	| tally browser tree tallyForClassMethodCall |
+	
+	tally := IPTTreeModelMock constructMessageTallyFrom: [IPTTreeModelMock computeSomething].
```
Comment (tillprochaska) [inline packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testBrowseClassMethod.st]: I could imagine that the test would be more robust if you used a `MessageTallyMock` here, as the assertion below is highly dependent on the implementation of the test itself.

File: packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testBrowseClassMethod.st [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+	tallyForClassMethodCall := tally receivers first.
+	
+	tree := IPTTreeModelMock new 
+		currentSelection: tallyForClassMethodCall.
+	browser := tree browseMethodCall.
+	
+	[self assert: browser selectedMessageName = #constructMessageTallyFrom:. 
+	self assert: browser selectedClassName = #IPTTreeModelMock] 
+		ensure: [browser dependents first delete]
```
Comment (tillprochaska) [inline packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testBrowseClassMethod.st]: Could you explain to me what this does? Tried to understand it, but I guess I’m lost 🙃

File: packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testCITests.st [hunk 13]
```diff
@@ -0,0 +1,9 @@
+tests
+testCITests
```
Comment (tillprochaska) [inline packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testCITests.st]: Diese Methode ist von eurem Test übrig geblieben, ob smalltalkCI fehlschlägt, selbst wenn man alle Fenster wieder schließt, oder?

File: packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testBrowseInstanceMethod.st [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+	
+	| browser tree |
+	
+	tree := IPTTreeModelMock new 
+		roots: {tallyObject};
+		currentSelection: tallyObject.
+		
+	browser := tree browseMethodCall.
+	
+	[self assert: browser selectedMessageName = #setUp.
```
Comment (tillprochaska) [inline packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testBrowseInstanceMethod.st]: I’d recommend using the shorthand `assert:equals:` here (and in the other assertions).

File: packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/explorerKey.from.event..st (hunk 3 above)
Comment (linusha) [reply packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/explorerKey.from.event..st]: The current way of implementing this is easier to extend if we were to add new keybindings in the future.

File: packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/explorerKey.from.event..st (hunk 4 above)
Comment (linusha) [reply packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/explorerKey.from.event..st]: This seems to be the standard way of doing it in Squeak. We copied this method from `Object Explorer`. We agree with you, but also think that consistency is important.

File: packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/mainTreeListMenu..st (hunk 5 above)
Comment (linusha) [reply packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/mainTreeListMenu..st]: We unfortunately cant. This is the way the `Object-Explorer` implements the menu.

File: packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/explorerKey.from.event..st (hunk 3 above)
Comment (Alexander-Dubrawski) [reply packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/explorerKey.from.event..st]: Ich finde besonders die folgende Variante die Till vorgeschlagen hat gut lesbar: 

```smaltalk 
(self currentSelection notNil) & (aChar == $b)
	ifTrue: [self browseMethodCall].
```
Hier ist ganz klar der Sinn zu erkennen.

File: packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/mainTreeListMenu..st (hunk 5 above)
Comment (Alexander-Dubrawski) [reply packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/mainTreeListMenu..st]: Ich verstehe die Syntax leider auch nicht :(

File: packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testBrowseClassMethod.st (hunk 12 above)
Comment (linusha) [reply packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testBrowseClassMethod.st]: Do you mean all three lines or just line 15? Line 15 closes any opened Browser windows at the end of the test. The two lines above test that the correct method (first line) and class (second line) are opened in the browser.

File: packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testCITests.st (hunk 13 above)
Comment (linusha) [reply packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testCITests.st]: Danke, wird entfernt.

File: packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/mainTreeListMenu..st (hunk 7 above)
Comment (linusha) [reply packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/mainTreeListMenu..st]: Same. We copied it for safety reasons. Quick test seemed okay without it. We removed it.

File: packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testBrowseClassMethod.st (hunk 11 above)
Comment (linusha) [reply packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testBrowseClassMethod.st]: We agree with you in principle, but it seems to us that this is currently not easily possible with the current Mock. We note this down for further refactorings in the future!

File: packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testBrowseClassMethod.st (hunk 12 above)
Comment (tillprochaska) [reply packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testBrowseClassMethod.st]: Thanks! In this case, why can’t we just have smth like the following?

```smalltalk
self assert: browser selectedMessageName = #constructMessageTallyFrom:. 
self assert: browser selectedClassName = #IPTTreeModelMock.
	
browser dependents first delete
```

File: packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/explorerKey.from.event..st (hunk 4 above)
Comment (tillprochaska) [reply packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/explorerKey.from.event..st]: Yes, I’ve seen the method in `ObjectExplorer` as well, but that seems to be the only place where it occurs. And in that context, I think it makes sense as the class is called explorer

Comment (twollnik) [reply packages/InteractiveProfilingTool-Core.package/IPTTreeModel.class/instance/explorerKey.from.event..st]: Ok, thanks for checking. I will change the message name.

File: packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testBrowseClassMethod.st (hunk 12 above)
Comment (twollnik) [reply packages/InteractiveProfilingTool-Tests.package/IPTTreeModelTestCase.class/instance/testBrowseClassMethod.st]: The `ensure` acts as a finally block that ensures that the browser window is always closed. 
If we do it like you suggested, if one of the asserts fails, the browser window will stay open.

**Pull request merged**